stubs/SQLAlchemy/sqlalchemy/engine/base.pyi

```diff
@@ -82,88 +118,115 @@ class ExceptionContextImpl(ExceptionContext):
+        execution_options: Mapping[str, Any] | None = ...,
         hide_parameters: bool = ...,
     ) -> None: ...
     @property
-    def engine(self): ...
+    def engine(self) -> Engine: ...
     def clear_compiled_cache(self) -> None: ...
     def update_execution_options(self, **opt) -> None: ...
     def execution_options(self, **opt): ...
     def get_execution_options(self): ...
     @property
-    def name(self): ...
+    def name(self) -> str: ...
     @property
     def driver(self): ...
     def dispose(self) -> None: ...
     class _trans_ctx:
-        conn: Any
-        transaction: Any
-        close_with_result: Any
-        def __init__(self, conn, transaction, close_with_result) -> None: ...
-        def __enter__(self): ...
-        def __exit__(self, type_, value, traceback) -> None: ...
-    def begin(self, close_with_result: bool = ...): ...
-    def transaction(self, callable_, *args, **kwargs): ...
-    def run_callable(self, callable_, *args, **kwargs): ...
-    def execute(self, statement, *multiparams, **params): ...
-    def scalar(self, statement, *multiparams, **params): ...
-    def connect(self, close_with_result: bool = ...): ...  # type: ignore[override]
-    def table_names(self, schema: Any | None = ..., connection: Any | None = ...): ...
-    def has_table(self, table_name, schema: Any | None = ...): ...
-    def raw_connection(self, _connection: Any | None = ...): ...
+        conn: Connection
+        transaction: Transaction
+        close_with_result: bool
+        def __init__(self, conn: Connection, transaction: Transaction, close_with_result: bool) -> None: ...
+        def __enter__(self) -> Connection: ...
+        def __exit__(
+            self, type_: type[BaseException] | None, value: BaseException | None, traceback: TracebackType | None
+        ) -> None: ...
+    def begin(self, close_with_result: bool = ...) -> _trans_ctx: ...
+    # TODO:
+    # def transaction(self, callable_: Callable[Concatenate[Connection, _P], _T], *args: _P.args, **kwargs: _P.kwargs) -> _T | None: ...
+    def transaction(self, callable_: Callable[..., _T], *args: Any, **kwargs: Any) -> _T | None: ...
+    # TODO:
+    # def run_callable(self, callable_: Callable[Concatenate[Connection, _P], _T], *args: _P.args, **kwargs: _P.kwargs) -> _T: ...
+    def run_callable(self, callable_: Callable[..., _T], *args: Any, **kwargs: Any) -> _T: ...
+    @overload  # type: ignore[override]
+    def execute(
+        self,
+        statement: ClauseElement | FunctionElement | DDLElement | DefaultGenerator | Compiled,
+        *multiparams: Mapping[str, Any],
+        **params: Any,
+    ) -> CursorResult: ...
+    @overload
+    def execute(self, statement: str, *multiparams: Any | tuple[Any, ...] | Mapping[str, Any], **params: Any) -> CursorResult: ...
+    @overload  # type: ignore[override]
+    def scalar(
+        self,
+        statement: ClauseElement | FunctionElement | DDLElement | DefaultGenerator | Compiled,
```

Comment:
- How did you figure out this type?
- The same type is repeated in `execute()`. Maybe we should make a type alias.